docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1060 @@
+
+As mentioned before, `Vector128<T>` and `Vector256<T>` are constrained to a specific set of primitive types. `char` is not one of them, but it does not mean that we can't implement vectorized text operations with the new APIs. For primitive types of the same size (and value types that don't contain references), casting is the solution.
+
+[Unsafe.As<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafe.as#system-runtime-compilerservices-unsafe-as-2(-0@)) can be used to get a reference to supported type:
+
+```cs
+void CastingReferences(Span<char> buffer)
+{
+    ref char charSearchSpace = ref MemoryMarshal.GetReference(buffer);
+    ref short searchSpace = ref Unsafe.As<char, short>(ref charSearchSpace);
+    // from now on we can use Vector128<short> or Vector256<short>
+}
+```
+
+Or [MemoryMarshal.Cast<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.memorymarshal.cast#system-runtime-interopservices-memorymarshal-cast-2(system-readonlyspan((-0)))), which casts a span of one primitive type to a span of another primitive type:
+
+```cs
+void CastingSpans(Span<char> chars)
+{
+    Span<short> shorts = MemoryMarshal.Cast<char, short>(chars);
+}
+```
+
+It's also possible to get managed references from unmanaged pointers:
+
+```cs
+void PointerToReference(char* pUtf16Buffer, byte* pAsciiBuffer)
+{
+    // of the same type:
+    ref byte asciiBuffer = ref *pAsciiBuffer;
+    // of different types:
+    ref ushort utf16Buffer = ref *(ushort*)pUtf16Buffer;
+}
+```
+
+We should avoid doing this in the opposite direction, as most engineers will assume that unmanaged pointers are already pinned.
+
+## Mindset
+
+Vectorizing real-world algorithms seems complex at the beginning. And what do software engineers do with complex problems? We break them down into sub-problems until these become simple enough to be solved directly.
+
+Let's implement a vectorized method for checking whether a given byte buffer consists only from valid ASCII characters to see how similar problems can be solved.
+
+### Edge cases
+
+Before we start working on the implementation, let's list all edge cases for our `IsAcii(ReadOnlySpan<byte> buffer)` method (and ideally write tests):
+
+* It does not need to throw any argument exceptions, as `ReadOnlySpan` is `struct` and it can never be `null` or invalid.
+* It should return `true` for an empty buffer.
+* It should detect invalid characters in the entire buffer, including the remainder.
+* It should not read any bytes that don't belong to the provided buffer.
+
+### Scalar solution
+
+Once we know all edge cases, we need to understand our problem and find a scalar solution.
+
+ASCII characters are values in the range from `0` to `127` (inclusive). It means that we can find invalid ASCII bytes by just searching for values that are larger than `127`. If we treat `byte` (unsigned) as `sbyte` (signed), it's a matter of performing "is less than zero" check.
```

Comment:
```suggestion
ASCII characters are values in the range from `0` to `127` (inclusive). It means that we can find invalid ASCII bytes by just searching for values that are larger than `127`. If we treat `byte` (unsigned, range from 0 to 255) as `sbyte` (signed, range from -128 to 127), it's a matter of performing "is less than zero" check.
```
Too verbose?
But it makes it obvious why the cast from `byte -> sbyte` is handy here.